README.md
# koa-ratelimit

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![node version][node-image]][node-url]


Rate limiter middleware for koa.


## Installation

```bash
# npm
$ npm install koa-ratelimit
# yarn
$ yarn add koa-ratelimit
```


## Example

### With a Redis driver

```js
const Koa = require('koa');
const ratelimit = require('koa-ratelimit');
const Redis = require('ioredis');
const app = new Koa();

// apply rate limit
app.use(ratelimit({
  driver: 'redis',
  db: new Redis(),
  duration: 60000,
  errorMessage: 'Sometimes You Just Have to Slow Down.',
  id: (ctx) => ctx.ip,
  headers: {
    remaining: 'Rate-Limit-Remaining',
    reset: 'Rate-Limit-Reset',
    total: 'Rate-Limit-Total'
  },
  max: 100,
  disableHeader: false,
  whitelist: (ctx) => {
    // some logic that returns a boolean
  },
  blacklist: (ctx) => {
    // some logic that returns a boolean
  }
}));

// response middleware
app.use(async (ctx) => {
  ctx.body = 'Stuff!';
});

// run server
app.listen(
  3000,
  () => console.log('listening on port 3000')
);
```

### With a memory driver

```js
const Koa = require('koa');
const ratelimit = require('koa-ratelimit');
const app = new Koa();

// apply rate limit
const db = new Map();

app.use(ratelimit({
  driver: 'memory',
  db: db,
  duration: 60000,
  errorMessage: 'Sometimes You Just Have to Slow Down.',
  id: (ctx) => ctx.ip,
  headers: {
    remaining: 'Rate-Limit-Remaining',
    reset: 'Rate-Limit-Reset',
    total: 'Rate-Limit-Total'
  },
  max: 100,
  disableHeader: false,
  whitelist: (ctx) => {
    // some logic that returns a boolean
  },
  blacklist: (ctx) => {
    // some logic that returns a boolean
  }
}));

// response middleware
app.use(async (ctx) => {
  ctx.body = 'Stuff!';
});

// run server
app.listen(
  3000,
  () => console.log('listening on port 3000')
);
```


## Options

  - `driver` memory or redis [redis]
  - `db` redis connection instance or Map instance (memory)
  - `duration` of limit in milliseconds [3600000]
  - `errorMessage` custom error message
  - `id` id to compare requests [ip]
  - `headers` custom header names
  - `max` max requests within `duration` [2500]
  - `disableHeader` set whether send the `remaining, reset, total` headers [false]
  - `remaining` remaining number of requests [`'X-RateLimit-Remaining'`]
  - `reset` reset timestamp [`'X-RateLimit-Reset'`]
  - `total` total number of requests [`'X-RateLimit-Limit'`]
  - `whitelist` if function returns true, middleware exits before limiting
  - `blacklist` if function returns true, `403` error is thrown
  - `throw` call ctx.throw if true


## Responses

Example 200 with header fields:

```
HTTP/1.1 200 OK
X-Powered-By: koa
X-RateLimit-Limit: 100
X-RateLimit-Remaining: 99
X-RateLimit-Reset: 1384377793
Content-Type: text/plain; charset=utf-8
Content-Length: 6
Date: Wed, 13 Nov 2013 21:22:13 GMT
Connection: keep-alive

Stuff!
```

Example 429 response:

```
HTTP/1.1 429 Too Many Requests
X-Powered-By: koa
X-RateLimit-Limit: 100
X-RateLimit-Remaining: 0
X-RateLimit-Reset: 1384377716
Content-Type: text/plain; charset=utf-8
Content-Length: 39
Retry-After: 7
Date: Wed, 13 Nov 2013 21:21:48 GMT
Connection: keep-alive

Rate limit exceeded, retry in 8 seconds
```


## License

[MIT](LICENSE)


##

[npm-image]: https://img.shields.io/npm/v/koa-ratelimit.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-ratelimit
[travis-image]: https://img.shields.io/travis/koajs/ratelimit.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/ratelimit
[node-image]: https://img.shields.io/badge/node.js-%3E=_10-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/


History.md

4.2.0 / 2019-02-03
==================

  * feat: add whitelist and blacklist opts (#44)

4.1.2 / 2018-06-13
==================

  * deps: remove unused dependence thenify (#37)

4.1.1 / 2018-05-29
==================

  * bump deps
  * throw correct error when rate limit exceeded (#36)

4.1.0 / 2018-04-08
==================

  * fix test error for custom error message
  * add opts.disableHeader support

4.0.0 / 2017-03-12
==================

  * Update signature to use async/await for koa@2

3.0.0 / 2017-03-12
==================

  * Update ratelimiter@3.0.2

2.4.0 / 2016-10-26
==================

  * Add custom error support

2.3.0 / 2016-05-12
==================

  * Add support for custom header names (#20)

2.2.0 / 2016-04-11
==================

  * Add option to throw error instead of setting body
  * bump deps

2.1.0 / 2015-01-04
==================

  * Merge pull request #11 from seegno/support/flush-redis-before-testing
  * Merge pull request #9 from seegno/enhancement/add-id-verification
  * Flush redis limit keys before running tests
  * Add id verification to bypass ratelimiting

2.0.0 / 2015-01-02
==================

  * Merge pull request #10 from seegno/support/update-ratelimiter-2-0-0
  * Update `debug@2.1.1` and `ms@0.7.0`
  * Update recommended node version to >= 0.11.13
  * Update dev dependencies
  * Update ratelimiter@2.0.0
  * Replace `thunkify` by `thenify`.

1.1.0 / 2014-12-03
==================

  * Adds callback for defining own comparator ID. Fixes #3

1.0.3 / 2014-05-24
==================

 * fix: don't yield to downstream middleware if ratelimit is hit.

1.0.2 / 2014-05-22
==================

 * yield to next only when rate limit isn't hit

1.0.1 / 2013-11-29
==================

 * fix ms dep


test/ratelimit.js

const Koa = require('koa');
const Redis = require('ioredis');
const request = require('supertest');
const should = require('should');

const ratelimit = require('..');
const db = new Redis();

describe('ratelimit middleware', () => {
  const duration = 1000;
  const goodBody = 'Num times hit:';

  beforeEach(async () => await db.eval(`for i, name in ipairs(redis.call('KEYS', 'limit:*')) do redis.call('DEL', name); end`, 0));

  describe('limit', () => {
    let guard;
    let app;

    const hitOnce = () => guard.should.equal(1);

    beforeEach(async () => {
      app = new Koa();

      app.use(ratelimit({ duration, db, max: 1 }));
      app.use(async (ctx) => {
        guard++;
        ctx.body = `${goodBody} ${guard}`;
      });

      guard = 0;

      await sleep(duration);
      await request(app.listen())
        .get('/')
        .expect(200, `${goodBody} 1`)
        .expect(hitOnce)
    });

    it('responds with 429 when rate limit is exceeded', async () => {
      await request(app.listen())
        .get('/')
        .expect('X-RateLimit-Remaining', '0')
        .expect(429);
    });

    it('should not yield downstream if ratelimit is exceeded', async () => {
      await request(app.listen())
        .get('/')
        .expect(429)

      hitOnce();
    });
  });

  describe('limit with throw', () => {
    let guard;
    let app;

    const hitOnce = () => guard.should.equal(1);

    beforeEach(async () => {
      app = new Koa();

      app.use(async (ctx, next) => {
        try {
          await next();
        } catch (e) {
          ctx.body = e.message;
          ctx.set(Object.assign({ 'X-Custom': 'foobar' }, e.headers));
        }
      });

      app.use(ratelimit({
        db,
        duration,
        max: 1,
        throw: true
      }));

      app.use(async (ctx) => {
        guard++;
        ctx.body = `${goodBody} ${guard}`;
      });

      guard = 0;

      await sleep(duration);
      await request(app.listen())
        .get('/')
        .expect(200, `${goodBody} 1`)
        .expect(hitOnce);
    });

    it('responds with 429 when rate limit is exceeded', async () => {
      await request(app.listen())
        .get('/')
        .expect('X-Custom', 'foobar')
        .expect('X-RateLimit-Remaining', '0')
        .expect((res) => res.error.text.should.match(/^Rate limit exceeded, retry in.*/))
        .expect(429)
    });
  });

  describe('id', async () => {
    it('should allow specifying a custom `id` function', async () => {
      const app = new Koa();

      app.use(ratelimit({
        db,
        id: (ctx) => ctx.request.header.foo,
        max: 1
      }));

      await request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect('X-RateLimit-Remaining', '0');
    });

    it('should not limit if `id` returns `false`', async () => {
      const app = new Koa();

      app.use(ratelimit({
        db,
        id: (ctx) => false,
        max: 5
      }));

      await request(app.listen())
        .get('/')
        .expect((res) => res.header.should.not.have.property('x-ratelimit-remaining'))
    });

    it('should limit using the `id` value', async () => {
      const app = new Koa();

      app.use(ratelimit({
        db,
        id: (ctx) => ctx.request.header.foo,
        max: 1
      }));

      app.use(async (ctx) => {
        ctx.body = ctx.request.header.foo;
      });

      await request(app.listen())
        .get('/')
        .set('foo', 'fiz')
        .expect(200, 'fiz');

      await request(app.listen())
        .get('/')
        .set('foo', 'biz')
        .expect(200, 'biz');
    });
  });

  describe('whitelist', () => {
    const duration = 1000;
    let guard;
    let app;

    const hitOnce = () => guard.should.equal(1);

    beforeEach(async () => {
      app = new Koa();

      app.use(ratelimit({
        db,
        whitelist: (ctx) => ctx.request.header.foo === 'whitelistme',
        max: 1
      }));
      app.use(async (ctx) => {
        guard++;
        ctx.body = 'foo';
      });

      guard = 0;

      await sleep(duration);
      await request(app.listen())
        .get('/')
        .expect(200)
        .expect(hitOnce)
    });

    it('should not limit if satisfy whitelist function', async () => {
      await request(app.listen())
        .get('/')
        .set('foo', 'whitelistme')
        .expect(200)

      await request(app.listen())
        .get('/')
        .set('foo', 'whitelistme')
        .expect(200)
    });

    it('should limit as usual if not whitelist return false', async () => {
      await request(app.listen())
        .get('/')
        .set('foo', 'imnotwhitelisted')
        .expect(429)
    });
  })

  describe('blacklist', () => {
    let app;

    beforeEach(async () => {
      app = new Koa();

      app.use(ratelimit({
        db,
        blacklist: (ctx) => ctx.request.header.foo === 'blacklisted',
        max: 1
      }));
      app.use(async (ctx) => {
        ctx.body = 'foo';
      });

    });

    it('should throw 403 if blacklisted', async () => {
      await request(app.listen())
        .get('/')
        .set('foo', 'blacklisted')
        .expect(403)
    });

    it('should return 200 when not blacklisted', async () => {
      await request(app.listen())
        .get('/')
        .set('foo', 'imnotblacklisted')
        .expect(200)
    });
  })

  describe('custom headers', () => {
    it('should allow specifying custom header names', async () => {
      const app = new Koa();

      app.use(ratelimit({
        db,
        headers: {
          remaining: 'Rate-Limit-Remaining',
          reset: 'Rate-Limit-Reset',
          total: 'Rate-Limit-Total'
        },
        max: 1
      }));

      await request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect((res) => {
          res.headers.should.have.keys('rate-limit-remaining', 'rate-limit-reset', 'rate-limit-total');
          res.headers.should.not.have.keys('x-ratelimit-limit', 'x-ratelimit-remaining', 'x-ratelimit-reset');
        });
    });
  });

  describe('custom error message', async () => {
    it('should allow specifying a custom error message', async () => {
      const app = new Koa();
      const errorMessage = 'Sometimes You Just Have to Slow Down.';

      app.use(ratelimit({
        db,
        errorMessage,
        max: 1
      }));

      app.use(async (ctx) => {
        ctx.body = 'foo';
      });

      await request(app.listen())
        .get('/')
        .expect(200);

      await request(app.listen())
        .get('/')
        .expect(429)
        .expect(errorMessage);
    });

    it('should return default error message when not specifying', async () => {
      const app = new Koa();

      app.use(ratelimit({
        db,
        max: 1,
      }));

      app.use(async (ctx) => {
        ctx.body = 'foo';
      });

      await request(app.listen())
        .get('/')
        .expect(200)

      await request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect(429)
        .expect((res) => res.text.should.match(/Rate limit exceeded, retry in \d+ minutes\./));
    });
  });

  describe('disable headers', () => {
    it('should disable headers when set opts.disableHeader', async () => {
      const app = new Koa();

      app.use(ratelimit({
        db,
        headers: {
          remaining: 'Rate-Limit-Remaining',
          reset: 'Rate-Limit-Reset',
          total: 'Rate-Limit-Total'
        },
        disableHeader: true,
        max: 1
      }));

      await request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect((res) => {
          res.headers.should.not.have.keys('rate-limit-remaining', 'rate-limit-reset', 'rate-limit-total');
          res.headers.should.not.have.keys('x-ratelimit-limit', 'x-ratelimit-remaining', 'x-ratelimit-reset');
        });
    });
  });
});

async function sleep(ms) {
 await new Promise((resolve, reject) => setTimeout(resolve, ms));
}


index.js

/**
 * Module dependencies.
 */

const debug = require('debug')('koa-ratelimit');
const Limiter = require('ratelimiter');
const ms = require('ms');

/**
 * Expose `ratelimit()`.
 */

module.exports = ratelimit;

/**
 * Initialize ratelimit middleware with the given `opts`:
 *
 * - `duration` limit duration in milliseconds [1 hour]
 * - `max` max requests per `id` [2500]
 * - `db` database connection
 * - `id` id to compare requests [ip]
 * - `headers` custom header names
 * - `remaining` remaining number of requests ['X-RateLimit-Remaining']
 * - `reset` reset timestamp ['X-RateLimit-Reset']
 * - `total` total number of requests ['X-RateLimit-Limit']
 * - `whitelist` whitelist function [false]
 * - `blacklist` blacklist function [false]
 *
 * @param {Object} opts
 * @return {Function}
 * @api public
 */

function ratelimit(opts = {}) {
  const {
    remaining = 'X-RateLimit-Remaining',
    reset = 'X-RateLimit-Reset',
    total = 'X-RateLimit-Limit'
  } = opts.headers || {}

  return async function ratelimit(ctx, next) {
    const id = opts.id ? opts.id(ctx) : ctx.ip;
    const whitelisted = typeof opts.whitelist === 'function' ? await opts.whitelist(ctx) : false;
    const blacklisted = typeof opts.blacklist === 'function' ? await opts.blacklist(ctx) : false;

    if (blacklisted) {
      ctx.throw(403, 'Forbidden')
    }

    if (false === id || whitelisted) return await next();

    // initialize limiter
    const limiter = new Limiter(Object.assign({}, opts, { id }));

    // check limit
    const limit = await thenify(limiter.get.bind(limiter));

    // check if current call is legit
    const calls = limit.remaining > 0 ? limit.remaining - 1 : 0;

    // check if header disabled
    const disableHeader = opts.disableHeader || false;

    let headers = {};
    if (!disableHeader) {
      // header fields
      headers = {
        [remaining]: calls,
        [reset]: limit.reset,
        [total]: limit.total
      };

      ctx.set(headers);
    }

    debug('remaining %s/%s %s', remaining, limit.total, id);
    if (limit.remaining) return await next();

    const delta = (limit.reset * 1000) - Date.now() | 0;
    const after = limit.reset - (Date.now() / 1000) | 0;
    ctx.set('Retry-After', after);

    ctx.status = 429;
    ctx.body = opts.errorMessage || `Rate limit exceeded, retry in ${ms(delta, { long: true })}.`;

    if (opts.throw) {
      ctx.throw(ctx.status, ctx.body, { headers });
    }
  }
}

/**
 * Helper function to convert a callback to a Promise.
 */

async function thenify(fn) {
  return await new Promise(function(resolve, reject) {
    function callback(err, res) {
      if (err) return reject(err);
      return resolve(res);
    }

    fn(callback);
  });
}
